### DISCUSSION TITLE
No preset with name "kAligner2".

### DISCUSSION THREAD

**Discussion by Angieiscute** [milaboratory/mixcr#1418] [Q&A] [answered]
Hi,

App version: 4.5.0;

When I employ the following command:
`mixcr align --rna --preset generic-bcr-amplicon --species hsa
`
A warning is raised:
```
WARNING: Preset generic-bcr-amplicon is deprecated. The preset is deprecated. Please use generic-amplicon.
```
Subsequently, I attempted to utilize:

`mixcr align --rna --preset generic-amplicon --preset kAligner2 --species hsa
`

This resulted in an error:
```
option '--preset' (<preset>) should be specified only once.
```

Hence, if I want to apply kAligner2 for BCR analysis, what are the correct options to use?

Thanks!

**Comment by mizraelson** [comment, marked as answer]
We have updated our `kAligner2` to accommodate both TCR and BCR data. As a result, we have replaced `generic-bcr-amplicon` and `generic-tcr-amplicon` with `generic-amplicon`. So there is no need to specify `kAligner2` separately.

**Reply by Angieiscute** [reply]
Thanks!

another question is about  --assemble-contigs-by,

`mixcr  assembleContigs --assemble-contigs-by VRegion  alignment.clna full.alignment.clns`

This resulted in an error:
Can't apply `--assemble-contigs-by VRegion` because `assembleContigs` step is not presented in the preset.

**Reply by mizraelson** [reply]
Yes, `generic-amplicon` was designed for targeted amplicon data which does not require `assembleContigs`. What kind of data do you work with?

**Reply by Angieiscute** [reply]
oh, we use 5RACE TCR data. The reason I want to use `--assemble-contigs-by`  is we would like to use `mixcr findAlleles` . 

```
mixcr4 findAlleles \
    --output-template full.findAlleles.clns \
    --export-alleles-mutations allele_stats.tsv \
    full.alignment.clns
```
This resulted in an error:
Some of the inputs were processed by assembleContigs without --assemble-contigs-by option

**Reply by mizraelson** [reply]
If you have 5RACE TCR there is no need to use `assembleContigs` function at all. Do you have UMIs in your data? Is the reverse primer in C gene? Do you reads overlap and are expected to cover the full VDJRegion?

**Reply by Angieiscute** [reply]
I would like to export full regions with uppercase and lowercase letters, so I need to have `assembleContigs` function. There is no UMI and primer is in C gene. And PE300 is used.

**Reply by mizraelson** [reply]
If you are using PE300 then most likely your reads should cover the full VDJRegion, or do you expect a gap?

**Reply by Angieiscute** [reply]
yes, there should be gaps in the V region which can not be covered.

**Reply by mizraelson** [reply]
Okay, so technically, what you want to achieve can be accomplished using the command below:

```
mixcr analyze generic-amplicon \
  --species hsa \
  --rna \
  --floating-left-alignment-boundary \
  --floating-right-alignment-boundary C \
  --add-step assembleContigs \
  --assemble-contigs-by VDJRegion \
  --impute-germline-on-export \
  input_R1.fastq.gz \
  input_R2.fastq.gz \
  result
```

However, when you use `--assemble-contigs-by VDJRegion`, it implies that only clones fully covering the VDJRegion will be exported. This means if all your sequences have a gap somewhere in the V gene, they will be filtered out. Therefore, I wouldn't recommend this approach.

If you omit `--assemble-contigs-by VDJRegion` from the command above, you should receive output wherein all the uncovered nucleotides are inferred from a germline but no clones will be filtered out. However, since this gap does not have a fixed position in all clones, these outputs would not be suitable for use with `mixcr findAlleles`.

The correct approach in your case would be to locate the ultimate borders of your gap and assemble clones by the feature that excludes that region. For instance, if your gap is always in CDR2, you can use the command below:

```
mixcr analyze generic-amplicon \
  --species hsa \
  --rna \
  --floating-left-alignment-boundary \
  --floating-right-alignment-boundary C \
  --assemble-clonotypes-by "[{FR1Begin:FR2End},{FR3Begin:FR4End}] \
  --impute-germline-on-export \
  input_R1.fastq.gz \
  input_R2.fastq.gz \
  result
```
Then, you can use these `.clns` files for `mixcr findAlleles`.

Let me know if that works.

**Reply by Angieiscute** [reply]
Thanks, BTW, how do I put '-OreadsLayout=Collinear' in the analyze?

**Reply by mizraelson** [reply]
What is the structure of your reads? R1 is always forward and R2 is always in reverse?

**Reply by Angieiscute** [reply]
yes, and we need to use Collinear in the alignment command line.

**Reply by Angieiscute** [reply]
oh, I recalled, it should be `-MreadsLayout=Collinear \`

**Reply by mizraelson** [reply]
`Collinear` is required if both of your reads are either forward or reverse. If your R1 is forward and R2 is reverse, you shouldn't use this parameter.

**Reply by Angieiscute** [reply]
I got u! Thanks!

**Reply by abspangler13** [reply]
Is this true for the smart-seq2-vdj preset? As in kAligner2 is the default.

**Reply by mizraelson** [reply]
Yes, `kALigner2` is now used in all presets.
